builder/api/src/main/java/io/helidon/builder/api/Prototype.java

```diff
@@ -0,0 +1,414 @@
+     */
+    interface ConfiguredBuilder<BUILDER, PROTOTYPE> extends Builder<BUILDER, PROTOTYPE> {
+        /**
+         * Update builder from configuration.
+         *
+         * @param config configuration to use
+         * @return updated builder instance
+         */
+        BUILDER config(Config config);
+
+        /**
+         * Discover services from configuration.
+         *
+         * @param config               configuration located at the node of the service providers
+         *                             (either a list node, or object, where each child is one service)
+         * @param serviceLoader        helidon service loader for the expected type
+         * @param allFromServiceLoader whether all services from service loader should be used, or only the ones with configured
+         *                             node
+         * @param <S>                  type of the expected service
+         * @return list of discovered services
+         */
+        default <S extends NamedService, T extends ConfiguredProvider<S>> List<S>
+        discoverServices(Config config,
+                         HelidonServiceLoader<T> serviceLoader,
+                         Class<T> providerType,
+                         Class<S> configType,
+                         boolean allFromServiceLoader) {
+            return ProvidedUtil.discoverServices(config, serviceLoader, providerType, configType, allFromServiceLoader);
+        }
+    }
+
+    /**
+     * A prototype {@link Prototype.Blueprint} may extend this interface
+     * to explicitly reference the associated runtime type.
+     * <p>
+     * For example a {@code RetryPrototypeBlueprint} that extends a {@link Prototype.Factory}
+     * of {@code Retry}, will add methods to the prototype interface and builder to build an instance of
+     * {@code Retry} directly.
+     * A factory method must exist either on the runtime type (such as {@code Retry}) with signature
+     * {@code static Retry create(RetryPrototype)}, or on the prototype blueprint.
+     *
+     * @param <T> type of the runtime object (such as {@code Retry} in the description above)
+     */
+    interface Factory<T> {
+        /**
+         * Create a new instance of the runtime type from this config object.
+         *
+         * @return new configured runtime instance
+         */
+        T build();
+        /*
+        This method intentionally clashes with io.helidon.common.Builder.build(), as we need to have it
+        available both on the builder and on the prototype instance.
+         */
+    }
+
+    /**
+     * A package local type (by design) that defines getter methods and possible static factory methods
+     * that form prototype information that is generated through annotation processing.
+     * <p>
+     * The following names are prohibited from being builder properties:
+     * <ul>
+     *     <li>{@code build} - builder terminating method that builds runtime type from builder,
```

Comment:
We should formalize this into a new `Reserved` type class.

Reply:
Is is in a single place in implementation - `TypeContext.IGNORED_NAMES`.
As the API is not used by the processor, there is no need to do this. Also there is not much use for a user, as this is about names you type into an interface, so such API would not bring benefits.
What use case do you envision here?

Reply:
I have also fixed this comment, as it is no longer fully correct.